src/igvfd/audit/sample.py

```diff
@@ -35,3 +35,53 @@ def audit_sample_sorted_fraction_parent_child_check(value, system):
         )
         if prop_errors != '':
             yield AuditFailure('inconsistent sorted fraction metadata', detail, level='ERROR')
+
+
+@audit_checker('Sample', frame='object')
+def audit_non_virtual_sample_linked_to_virtual_sample(value, system):
+    '''Non-virtual samples should not be linked to virtual samples'''
+    sample_id = system.get('path')
+    sample_is_virtual = value.get('virtual', False)
+    links_to_check = [
+        value.get('part_of', None),
+        value.get('originated_from', None),
+        value.get('sorted_fraction', None),
+    ]
+    links_to_check.extend(value.get('pooled_from', []))
+    for linked_sample in links_to_check:
+        audit_failure = get_virtual_sample_failures(
+            system,
+            sample_id,
+            sample_is_virtual,
+            linked_sample
+        )
+        if audit_failure:
+            yield audit_failure
+
+
+def get_virtual_sample_failures(
+    system,
+    sample_id,
+    sample_is_virtual,
+    linked_sample_id
+):
+    if not linked_sample_id:
+        return None
```

Comment:
this will never happen, you can remove

Reply:
This change causes the unit test to fail.  The lines on 46 to 48 potentially return None in the list of linked IDs if the property is not present in the sample.

Reply:
if the linked IDs are not present then links_to_check should be empty list?

Reply:
But we are populating them with value None if the property is not present. Each entry will be an element of the list.

Reply:
I tried the unit tests with these lines removed, and the test fails.